File: lintcode/python/0141_sqrt_x.py
#!/usr/bin/python -t

# binary search solution, time O(logn) space O(1)

class Solution:
    """
    @param x: An integer
    @return: The sqrt of x
    """
    def sqrt(self, x):
        # write your code here
        if x == 0:
            return 0
        if x == 1:
            return 1
            
        l = 0
        r = x
        
        while l < r:
            mid = (l+r)/2
            if mid*mid == x:
                return mid
            elif mid*mid < x:
                l = mid+1
            else:
                r = mid
                
        return l-1

File: lintcode/golang/0141_sqrt_x.go
/**
 * @param x: An integer
 * @return: The sqrt of x
 */
 
//import "fmt"

func sqrt (x int) int {
    // write your code here
    if x <= 1 {
        return x
    }
    l := 1
    r := x-1
    
    for l + 1 < r {
        mid := (l+r)/2
        val := mid * mid
        //fmt.Println(mid, val)
        
        if val == x {
            return mid
        } else if val > x {
            r = mid
        } else {
            l = mid
        }
    }
    
    //fmt.Println(l, r)
    if r * r < x {
        return r
    }
    return l
}

